File: services/ingestion/infrastructure/adapters/postman_parser.py
"""Postman collection parser adapter."""
import logging
from typing import Dict, Any, List

logger = logging.getLogger(__name__)


class PostmanParser:
    """Adapter for parsing Postman collections into our API configuration format."""
    
    async def parse(self, collection_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Parse Postman collection to our API configuration format.
        
        Args:
            collection_data: Postman collection JSON
            
        Returns:
            API configuration dictionary
            
        Raises:
            ValueError: If collection is invalid
        """
        try:
            # Extract collection info
            info = collection_data.get('info', {})
            name = info.get('name', 'Unnamed API')
            description = info.get('description', '')
            version = info.get('version', '1.0.0')
            
            # Parse items (requests)
            items = collection_data.get('item', [])
            endpoints = []
            
            for item in items:
                endpoints.extend(self._parse_item(item))
            
            # Build our configuration format
            config = {
                'name': name,
                'version': version,
                'description': description,
                'endpoints': endpoints,
                'source': 'postman',
                'postman_id': info.get('_postman_id'),
            }
            
            logger.info(f"Parsed Postman collection: {name} with {len(endpoints)} endpoints")
            return config
            
        except Exception as e:
            logger.error(f"Error parsing Postman collection: {str(e)}")
            raise ValueError(f"Invalid Postman collection: {str(e)}")
    
    def _parse_item(self, item: Dict[str, Any], prefix: str = '') -> List[Dict[str, Any]]:
        """
        Recursively parse Postman items (can be requests or folders).
        
        Args:
            item: Postman item
            prefix: Path prefix for nested items
            
        Returns:
            List of endpoint dictionaries
        """
        endpoints = []
        
        # Check if this is a folder (has nested items)
        if 'item' in item:
            folder_name = item.get('name', '')
            new_prefix = f"{prefix}/{folder_name}" if prefix else folder_name
            
            for nested_item in item['item']:
                endpoints.extend(self._parse_item(nested_item, new_prefix))
        
        # Check if this is a request
        elif 'request' in item:
            endpoint = self._parse_request(item, prefix)
            if endpoint:
                endpoints.append(endpoint)
        
        return endpoints
    
    def _parse_request(self, item: Dict[str, Any], prefix: str) -> Dict[str, Any]:
        """
        Parse a Postman request item.
        
        Args:
            item: Postman request item
            prefix: Path prefix
            
        Returns:
            Endpoint dictionary
        """
        request = item['request']
        
        # Extract method
        method = request.get('method', 'GET').upper()
        
        # Extract URL
        url = request.get('url', {})
        if isinstance(url, str):
            path = url
        elif isinstance(url, dict):
            raw_url = url.get('raw', '')
            # Try to extract path from raw URL
            path = self._extract_path_from_url(raw_url)
        else:
            path = ''
        
        # Add prefix
        if prefix:
            path = f"/{prefix}{path}" if not path.startswith('/') else f"/{prefix}{path}"
        
        # Extract description
        description = item.get('name', '')
        if 'description' in item:
            desc_obj = item['description']
            if isinstance(desc_obj, str):
                description = desc_obj
            elif isinstance(desc_obj, dict):
                description = desc_obj.get('content', description)
        
        # Extract headers
        headers = {}
        if 'header' in request:
            for header in request['header']:
                if not header.get('disabled', False):
                    key = header.get('key', '')
                    value = header.get('value', '')
                    if key:
                        headers[key] = value
        
        # Extract body
        body_schema = None
        if 'body' in request:
            body = request['body']
            mode = body.get('mode', '')
            
            if mode == 'raw':
                # Try to infer schema from raw body
                raw = body.get('raw', '')
                body_schema = self._infer_schema_from_example(raw)
        
        # Extract parameters from URL
        parameters = []
        if isinstance(url, dict) and 'variable' in url:
            for var in url['variable']:
                parameters.append({
                    'name': var.get('key', ''),
                    'type': 'string',
                    'description': var.get('description', ''),
                    'required': not var.get('disabled', False),
                })
        
        endpoint = {
            'name': item.get('name', ''),
            'path': path,
            'method': method,
            'description': description,
            'headers': headers,
            'parameters': parameters,
        }
        
        if body_schema:
            endpoint['request_schema'] = body_schema
        
        return endpoint
    
    def _extract_path_from_url(self, url: str) -> str:
        """Extract path from full URL."""
        # Remove protocol and domain
        if '://' in url:
            url = url.split('://', 1)[1]
        if '/' in url:
            path = '/' + url.split('/', 1)[1]
        else:
            path = '/'
        
        # Remove query string
        if '?' in path:
            path = path.split('?', 1)[0]
        
        return path
    
    def _infer_schema_from_example(self, example: str) -> Dict[str, Any]:
        """Try to infer JSON schema from example data."""
        try:
            import json
            data = json.loads(example)
            
            # Simple schema inference
            schema = {
                'type': 'object',
                'properties': {}
            }
            
            if isinstance(data, dict):
                for key, value in data.items():
                    schema['properties'][key] = {
                        'type': self._get_json_type(value)
                    }
            
            return schema
        except:
            return {'type': 'string'}
    
    def _get_json_type(self, value: Any) -> str:
        """Get JSON schema type from Python value."""
        if isinstance(value, bool):
            return 'boolean'
        elif isinstance(value, int):
            return 'integer'
        elif isinstance(value, float):
            return 'number'
        elif isinstance(value, str):
            return 'string'
        elif isinstance(value, list):
            return 'array'
        elif isinstance(value, dict):
            return 'object'
        else:
            return 'string'

File: services/ingestion/adapters/postman_parser.go
package adapters

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/testpilot-ai/ingestion/domain/entities"
)

// PostmanParser handles parsing of Postman collections
type PostmanParser struct{}

// NewPostmanParser creates a new Postman parser
func NewPostmanParser() *PostmanParser {
	return &PostmanParser{}
}

// ParseCollection parses a Postman collection file
func (p *PostmanParser) ParseCollection(filePath string) (*entities.APIConfig, string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, "", fmt.Errorf("failed to read file: %w", err)
	}

	return p.ParseCollectionData(data)
}

// ParseCollectionData parses Postman collection from bytes
func (p *PostmanParser) ParseCollectionData(data []byte) (*entities.APIConfig, string, error) {
	var collection entities.PostmanCollection
	if err := json.Unmarshal(data, &collection); err != nil {
		return nil, "", fmt.Errorf("failed to parse Postman collection: %w", err)
	}

	// Calculate hash
	fileParser := NewFileParser()
	contentHash := fileParser.CalculateHash(data)

	// Extract base_url from the first request's host
	baseURL := p.extractBaseURL(collection.Item)

	// Convert to APIConfig
	config := &entities.APIConfig{
		Name:        collection.Info.Name,
		Version:     "1.0.0",
		Description: collection.Info.Description,
		BaseURL:     baseURL,
		Endpoints:   p.extractEndpoints(collection.Item),
	}

	return config, contentHash, nil
}

// extractBaseURL extracts the base URL from the first request in the collection
func (p *PostmanParser) extractBaseURL(items []entities.PostmanItem) string {
	for _, item := range items {
		// If it's a folder, recurse
		if len(item.Item) > 0 {
			if baseURL := p.extractBaseURL(item.Item); baseURL != "" {
				return baseURL
			}
			continue
		}

		// If it has a request with a host, extract base URL
		if item.Request != nil && len(item.Request.URL.Host) > 0 {
			host := strings.Join(item.Request.URL.Host, ".")
			// Check if it looks like a variable (e.g., {{base_url}})
			if strings.Contains(host, "{{") {
				// Try to get from raw URL
				if item.Request.URL.Raw != "" {
					// Parse raw URL to extract host
					raw := item.Request.URL.Raw
					// Remove variable syntax and try to find a real host
					if strings.HasPrefix(raw, "http://") || strings.HasPrefix(raw, "https://") {
						// Extract host from raw URL
						parts := strings.SplitN(raw, "/", 4)
						if len(parts) >= 3 {
							return parts[0] + "//" + parts[2]
						}
					}
				}
				continue
			}
			// Build proper base URL
			protocol := "http://"
			if strings.Contains(item.Request.URL.Raw, "https://") {
				protocol = "https://"
			}
			return protocol + host
		}
	}
	return ""
}

// extractEndpoints recursively extracts endpoints from Postman items
func (p *PostmanParser) extractEndpoints(items []entities.PostmanItem) []entities.APIEndpoint {
	var endpoints []entities.APIEndpoint

	for _, item := range items {
		// If it's a folder, recurse (do NOT prepend folder names to paths)
		if len(item.Item) > 0 {
			endpoints = append(endpoints, p.extractEndpoints(item.Item)...)
			continue
		}

		// If it has a request, convert it
		if item.Request != nil {
			endpoint := p.convertRequest(item)
			endpoints = append(endpoints, endpoint)
		}
	}

	return endpoints
}

// convertRequest converts a Postman request to an APIEndpoint
func (p *PostmanParser) convertRequest(item entities.PostmanItem) entities.APIEndpoint {
	req := item.Request

	// Build path from URL (just the path, no folder prefixes)
	path := "/" + strings.Join(req.URL.Path, "/")

	endpoint := entities.APIEndpoint{
		Name:        item.Name,
		Path:        path,
		Method:      req.Method,
		Description: item.Name,
		Parameters:  p.extractParameters(req),
	}

	// Extract auth from headers
	for _, header := range req.Header {
		if strings.ToLower(header.Key) == "authorization" ||
			strings.ToLower(header.Key) == "x-api-key" {
			endpoint.Authentication = &entities.AuthConfig{
				Type:   "api_key",
				Header: header.Key,
			}
			break
		}
	}

	// Parse request body if present - preserve proper types
	if req.Body != nil && req.Body.Mode == "raw" && req.Body.Raw != "" {
		var bodySchema map[string]interface{}
		if err := json.Unmarshal([]byte(req.Body.Raw), &bodySchema); err == nil {
			endpoint.RequestSchema = bodySchema
		}
	}

	return endpoint
}

// extractParameters extracts parameters from a Postman request
func (p *PostmanParser) extractParameters(req *entities.PostmanRequest) []entities.Parameter {
	var params []entities.Parameter

	// Extract query parameters
	for _, q := range req.URL.Query {
		params = append(params, entities.Parameter{
			Name:     q.Key,
			Type:     "string",
			Required: false,
		})
	}

	return params
}
